teps/0015-implicit-mapping.md

```diff
@@ -0,0 +1,658 @@
+I.e., a `Pipeline` author will be able to omit parameters and resources for a `PipelineTask`.
+As a result, the previously listed *deploy* `Task` could be specified as followed in a `Pipeline` specification:
+
+```yaml
+    - name: deploy
+      runAfter: [verify]
+      taskRef:
+        name: install-tekton-release
+```
+
+### Goal
+- Simplify the `Pipeline` specification and therefore the `Pipeline` YAML by:
+  - Allowing a `Pipeline` author to omit parameters and resources from `PipelineTasks` that match in name with parameters and resources listed under the `Pipeline.spec.params` and `Pipeline.spec.resources` fields.
+
+* Support finally `Tasks` for parameters and resources (no support for using results from previous `Tasks`)
+
+### Non Goals
+- Support implicit mapping for parameters and resources in `Conditions`.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+|#|Requirement                                                                                                                                                                                                                                                                                                                                                                          |
+|-|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+|1| A `Pipeline` author can omit the explicit specification of parameters and resources for some, or any, `Task` in the `Pipeline.spec.tasks` field if:<br><br> <ul><li> The pipeline is annotated for implicit mapping <i><br>AND<br></i> <li> The omitted parameters and resources can be resolved according to Req. 2 </li></ul>|
+|2| Tekton tries to resolve omitted parameters and resources for a pipeline `Task` in the following order: <br><br> <ol><li> If any preceding `Task` has a `Task` result whose name matches an implicit parameter name of the `Task`, resolve the implicit parameter value with the last produced `Task` result that matches. <li> If a parameter or resource name in the `Pipeline.spec` field matches a `Task's` implicit parameter or resource name, resolve the parameter and/or resource values. </li> <li> Tekton produces an error if parameters or resources can't be matched. <br> *\*Explicitly specified params and resources always take precedence.*
+
+[DISCUSSION]: Additionally, omitting input resources that use the output resource of a previous task by using the [`from`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-the-from-parameter) parameter could be supported.
```

Comment:
To add even more possible features... we could implicitly map named workspaces in a Pipeline to named workspaces in Tasks:

```yaml
# PipelineSpec
spec:
  workspaces:
  - name: source
  - name: output
  tasks:
  - taskRef: name: lint
    workspaces:
    - name: source
      workspace: source
  - taskRef: name: unit-test
    workspaces:
    - name: source
      workspace: source
  - taskRef: name: build
    workspaces:
    - name: source
      workspace: source
    - name: output
      workspace: output
  # below just shows one without a matching workspace name for reference
  - taskRef: name: integration-test
    workspaces:
    - name: binary-to-test
      workspace: output
```

Would become...

```yaml
# PipelineSpec
spec:
  workspaces:
  - name: source
  - name: output
  tasks:
  - taskRef:
      name: lint
  - taskRef:
      name: unit-test
  - taskRef:
      name: build
  - taskRef:
      name: integration-test
    workspaces:
    - name: binary-to-test
      workspace: output
```